posts/2024-12-12-non-api-use/index.qmd

```diff
@@ -0,0 +1,1276 @@
+data that happens during normal R computations.) Until version 4.0.0,
+`NAMED` was R's mechanism to save memory and CPU time instead of
+creating and storing these copies. A temporary object such as the value
+of `1:10` was not bound to a symbol and thus could be modified right
+away. Assigning it to a variable, as in `x <- 1:10`, gave it a
+`NAMED(x)` count of 1, for which R had an internal optimisation in
+replacement function calls like `foo(x) <- 3`.  Assigning the same value
+to yet another symbol (by copying `y <- x` or calling a function
+`foo(x)`) increased the `NAMED()` count to 2 or more, for which there
+was no optimisation: in order to modify one of the symbols, R was
+required to duplicate `x` first. `NAMED()` was not necessarily decreased
+after the bindings disappeared, and decreasing it after having reached
+`NAMEDMAX` was impossible. During the lifetime of R-3.x, `NAMEDMAX` was
+increased from 2 to 3 and later to 7.
+
+Between R-3.1.0 and R-4.0.0, R [migrated from `NAMED` to reference
+counting][Tierney_refcnt]. Reference counts are easier to properly
+decrement than `NAMED`, thus preventing unneeded copies of objects that
+became unreferenced. R-3.5.0 [documented the symbols][Rnews_setnamed]
+`MAYBE_REFERENCED(.)` / `NO_REFERENCES(.)` for use instead of checking
+`NAMED(.) == 0`, `MAYBE_SHARED(.)` / `NOT_SHARED(.)` instead of checking
+`NAMED(.) > 1`, and `MARK_NOT_MUTABLE(.)` instead of setting `NAMED(.)`
+to `NAMEDMAX`, which later became part of the API instead of the
+`NAMED(.)` and `REFCNT(.)` functions. The hard rules are that a value is
+safe to modify in place if it has `NO_REFERENCES()` (reference count of
+0), definitely unsafe to modify in place (requiring a call to
+`duplicate` or `shallow_duplicate`) if it is `MAYBE_SHARED()` (reference
+count above 1), and almost certainly unsafe to modify in place if it is
+`MAYBE_REFERENCED()` (reference count of 1).
+
+`data.table`'s only uses of `NAMED()` were in the [verbose output during
+assignment][remove_named]:
+
+```c
+if (verbose) {
+  Rprintf(_("RHS for item %d has been duplicated because NAMED==%d MAYBE_SHARED==%d, but then is being plonked. length(values)==%d; length(cols)==%d)\n"),
+          i+1, NAMED(thisvalue), MAYBE_SHARED(thisvalue), length(values), length(cols));
+            // ^^^^^ non-API function
+}
+```
+
+Since the correctness of the modification operation hinges on the
+reference count being 0 (and it may be important whether it's exactly 1
+or above 1), the same amount of _useful_ information can be conveyed by
+printing `MAYBE_REFERENCED()` and `MAYBE_SHARED()` instead of `NAMED()`:
+
+```c
+if (verbose) {
+  Rprintf(_("RHS for item %d has been duplicated because MAYBE_REFERENCED==%d MAYBE_SHARED==%d, but then is being plonked. length(values)==%d; length(cols)==%d)\n"),
+          i+1, MAYBE_REFERENCED(thisvalue), MAYBE_SHARED(thisvalue), length(values), length(cols));
+            // ^^^^^^^^^^^^^^^^ API function
```

Comment:
is there a PR for this suggested change?

I see a link to a PR for another issue above, so I assume there is none for this one? so should somebody submit a PR?

more generally, it would be useful to have a comment/sentence at the end about the current status of each issue, so we can know if more work still needs to be done.